masq/src/line_reader.rs

```diff
@@ -1,242 +1,300 @@
-        LineReader {
-            delegate: Box::new(EditorReal::default()),
+    fn read_line(&self) -> TerminalEvent {
+        match self.interface.read_line() {
+            Ok(ReadResult::Input(line)) => {
+                self.add_history_unique(line.clone());
+                TerminalEvent::CommandLine(line)
+            }
+            Err(e) => TerminalEvent::Error(format!("Reading from the terminal: {}", e)),
+            Ok(ReadResult::Signal(Signal::Resize)) | Ok(ReadResult::Signal(Signal::Continue)) => {
+                TerminalEvent::Continue
+            }
+            _ => TerminalEvent::Break,
         }
     }
+
+    fn add_history_unique(&self, line: String) {
+        self.interface.add_history_unique(line)
+    }
+
+    ////////////////////////////////////////////////////////////////////////////////////////////////////
+
+    #[cfg(test)]
+    fn tell_me_who_you_are(&self) -> String {
+        format!(
+            "TerminalReal<{}>",
+            self.interface
+                .lock_writer_append()
+                .unwrap()
+                .tell_me_who_you_are()
+        )
+    }
 }
 
-trait EditorTrait {
-    fn readline(&mut self, prompt: &str) -> Result<String, ReadlineError>;
-    fn add_history_entry(&mut self, line: &str) -> bool;
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+//utils for integration tests run in the interactive mode
+
+pub struct IntegrationTestTerminal {
+    lock: Arc<Mutex<()>>,
+    stdin: Arc<Mutex<Box<dyn Read + Send + 'static>>>,
+    stdout: Arc<Mutex<Box<dyn Write + Send + 'static>>>,
 }
 
-struct EditorReal {
-    delegate: Editor<()>,
+impl Default for IntegrationTestTerminal {
+    fn default() -> Self {
+        IntegrationTestTerminal {
+            lock: Arc::new(Mutex::new(())),
+            stdin: Arc::new(Mutex::new(Box::new(stdin()))),
+            stdout: Arc::new(Mutex::new(Box::new(stdout()))),
+        }
+    }
 }
 
-impl EditorTrait for EditorReal {
-    fn readline(&mut self, prompt: &str) -> Result<String, ReadlineError> {
-        self.delegate.readline(prompt)
+impl MasqTerminal for IntegrationTestTerminal {
+    fn provide_lock(&self) -> Box<dyn WriterLock + '_> {
+        Box::new(IntegrationTestWriter {
+            temporary_mutex_guard: self.lock.lock().expect("providing MutexGuard failed"),
+        })
     }
 
-    fn add_history_entry(&mut self, line: &str) -> bool {
-        self.delegate.add_history_entry(line)
+    fn read_line(&self) -> TerminalEvent {
+        let mut buffer = [0; 1024];
+        let number_of_bytes = self
+            .stdin
+            .lock()
+            .expect("poisoned mutex")
+            .read(&mut buffer)
+            .expect("reading failed");
+        let _lock = self
+            .lock
+            .lock()
+            .expect("poisoned mutex in IntegrationTestTerminal");
+        short_writeln!(self.stdout.lock().unwrap(), "{}", MASQ_PROMPT);
```

Comment:
I see an `unwrap()` in production code! True, it's not _really_ production code, but it looks enough like production code to be distracting.